### PR TITLE
Issues/499 fitness should be float

### PR DESCRIPTION
**Pull request by jessesnyder**

### CODE REVIEW

**Comment by jessesnyder** [comment]
While trying to assess the implications of making the change to the `fitness` expression, I went looking for where it was used, and found the DiscreteGenerational network used by the Rogers demo. This network had a complex-looking commented out test (so far, so good!) and the access of the fitness attribute was pretty deep in the `add_node` method. After a while I figured out that the expression in question isn't actually touched via python (it's something to do with SQLA queries, and I'll need to test that separately), but I got embroiled in writing tests for the DiscreteGenerational network and found a couple bugs.

The first is just documentation, but it cost me some time, so it's worth mentioning:
The docstring for the `add_node` method ("Link the agent to a random member of the previous generation.") is sort of accurate, though I think it's more complicated than that. The docstring for the class is definitely wrong and misleading:

> "Each agent is connected to *all* agents in the previous generation." 

(my emphasis)

The original commented-out test was testing the behavior described in the (incorrect) class docstring, and I first re-wrote the test using the same assumption. Debugging ensued, and then I saw the method docstring, etc. etc.

The second bug has to do with marshaling back and forth between strings and bools in the `initial_source` property. The way it's done doesn't work, and causes problems for code which consults that attribute. There are now two failing tests related to that. I don't want to just go and fix it, since I don't know if anything is depending on the bug (like SimCity on Windows 95!)

**Comment by thomasmorgan** [comment]
@jessesnyder 

yes, I wrote the expression code, I don't fully understand the details of it, but having a getter allows you to access the values with python, e.g. `object.value` while the expression allows you to query by it, e.g. `ObjectClass.query.filter_by(value=x).all()`

You are right that the doc string is definitely wrong. That's how the discrete generational network used to behave a very long time ago. It was before we implemented indexing on the tables and so the mass creation of vectors caused tremendous slow down. To fix this (in addition to indexing) we made it so you only connect to nodes on an as-needed basis.

You are also correct that the connection method is a little more complicated than just "random". It would be more accurate to say that "a parent is selected with probability proportional to their fitness and they connect to you". This is how natural selection occurs via this network: fitter nodes tend to have more children.

In this case I suggest we fix the tests and documentation to reflect the actual behavior of the network.

However, I think we should fix the second bug. This network is used a lot, but I'd rather see it fixed and have to update experiments that use it than leave it broken. The initial_source variable isn't used much though so hopefully nothing will break! Happy to review fixes to the code.

**Comment by suchow** [comment]
@jessesnyder I will defer to @thomasmorgan on this because I'm not as familiar with the code. I'll note only that this PR doesn't include the actual change to the fitness getter, but I assume that's because you're figuring out the right way to fix it before making the fix.

**Comment by jessesnyder** [comment]
Thanks @thomasmorgan and @suchow. Fixed bug and docs, so new tests pass. I'll get back on the actual task (the sqla expression) asap!

File: dallinger/networks.py [hunk 1]
```diff
@@ -134,10 +139,10 @@ def generation_size(self):
     @property
     def initial_source(self):
         """The source that seeds the first generation."""
-        return bool(self.property3)
+        return self.property3.lower() != 'false'
 
     def add_node(self, node):
-        """Link the agent to a random member of the previous generation."""
+        """Link to the agent from a parent based on the paren'ts fitness"""
```
Comment (suchow) [inline dallinger/networks.py]: typo: `paren'ts` => `parent's`

**Comment by jessesnyder** [comment]
@suchow Actual bug now fixed.

@thomasmorgan I'm going to postpone any change to how numeric values are stored in the generic `propertyX` attributes, since they're all over the place and if they're going to be changed, they should probably all be changed at once. As @MatthewWilkes pointed out, there is a difference in precision between `str()` and `repr()`, with `repr()` retaining more precision:
```
(Pdb) str(1.0 / 3.0)
'0.333333333333'
(Pdb) repr(1.0 / 3.0)
'0.3333333333333333'
```

We could standardize the precision using a format string:
```
>>> "{:.6f}".format(1.0 / 3.0)
'0.333333'
>>> "{:.6f}".format(1.0 / 2.0)
'0.500000'
```

I can imagine that for querying, more precision isn't necessarily better.

**Pull request merged**